# Chosing, exploring and wrangling data set - Content list

1. Introduction
2. Import libraries and data
3. Explore and clean TGV data set
4. Explore and clean Number of travelers data set

# 1. Introduction

For this project, I decided to work with open data from SNCF, France's national state-owned railway company. All the data I will use in this project is open and was found on their website https://ressources.data.sncf.com/. I will combine several data sets in order to find relationships between different variables, such as the delays experienced in the train station, the number of passengers per year in a given train station, the region from which the train left, amongst others.

# 2. Import libraries and data

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

import warnings
warnings.filterwarnings('ignore')

In [2]:
# Create path
path = r'C:\Users\Mathilde\Documents\DATA ANALYSIS CAREERFOUNDRY\Aug 2024 - SNCF project'
# Data set on the delays on the TGV (high speed trains)
df_delay_tgv = pd.read_csv(os.path.join(path, '02 Data', 'Original data', 'regularite-mensuelle-tgv-aqst.csv'), index_col = False, delimiter=';')
#  Data set on the number of passengers in the train stations
df_nb_travelers = pd.read_csv(os.path.join(path, '02 Data', 'Original data', 'frequentation-gares.csv'), index_col = False, delimiter=';')

# 3. Explore and clean TGV data set

In [3]:
pd.set_option('display.max_columns', None)
df_delay_tgv.head()

Date   Service      Gare de départ       Gare d'arrivée  \
0  2018-01  National    BORDEAUX ST JEAN   PARIS MONTPARNASSE   
1  2018-01  National   LA ROCHELLE VILLE   PARIS MONTPARNASSE   
2  2018-01  National  PARIS MONTPARNASSE              QUIMPER   
3  2018-01  National  PARIS MONTPARNASSE              ST MALO   
4  2018-01  National  PARIS MONTPARNASSE  ST PIERRE DES CORPS   

   Durée moyenne du trajet  Nombre de circulations prévues  \
0                      141                             870   
1                      165                             222   
2                      220                             248   
3                      156                             102   
4                       61                             391   

   Nombre de trains annulés  Commentaire annulations  \
0                         5                      NaN   
1                         0                      NaN   
2                         1                      NaN   
3                         0                      NaN   
4                         2                      NaN   

   Nombre de trains en retard au départ  \
0                                   289   
1                                     8   
2                                    37   
3                                    12   
4                                    61   

   Retard moyen des trains en retard au départ  \
0                                    11.247809   
1                                     2.875000   
2                                     9.501351   
3                                    19.912500   
4                                     7.796995   

   Retard moyen de tous les trains au départ  Commentaire retards au départ  \
0                                   3.693179                            NaN   
1                                   0.095796                            NaN   
2                                   1.003981                            NaN   
3                                   1.966667                            NaN   
4                                   0.886889                            NaN   

   Nombre de trains en retard à l'arrivée  \
0                                     147   
1                                      34   
2                                      26   
3                                      13   
4                                      71   

   Retard moyen des trains en retard à l'arrivée  \
0                                      28.436735   
1                                      21.524020   
2                                      55.692308   
3                                      48.623077   
4                                      12.405164   

   Retard moyen de tous les trains à l'arrivée  \
0                                     6.511118   
1                                     5.696096   
2                                     7.578947   
3                                     6.790686   
4                                     3.346487   

                     Commentaire retards à l'arrivée  \
0                                                NaN   
1                                                NaN   
2  Ce mois-ci, l'OD a été touchée par les inciden...   
3  Ce mois-ci, l'OD a été touchée par les inciden...   
4                                                NaN   

   Nombre trains en retard > 15min  \
0                              110   
1                               22   
2                               26   
3                                8   
4                               17   

   Retard moyen trains en retard > 15 (si liaison concurrencée par vol)  \
0                                           6.511118                      
1                                           5.696096                      
2                                           7.548387                      
3                                           6.724757                      
4                                           

In [4]:
df_delay_tgv.shape

(9598, 26)

In [5]:
df_delay_tgv.describe()

Durée moyenne du trajet  Nombre de circulations prévues  \
count              9598.000000                     9598.000000   
mean                170.010940                      265.281413   
std                  87.300663                      178.879185   
min                   0.000000                        0.000000   
25%                 100.000000                      147.000000   
50%                 163.000000                      227.000000   
75%                 222.000000                      346.000000   
max                 786.000000                     1100.000000   

       Nombre de trains annulés  Commentaire annulations  \
count               9598.000000                      0.0   
mean                  10.070744                      NaN   
std                   24.784164                      NaN   
min                    0.000000                      NaN   
25%                    0.000000                      NaN   
50%                    2.000000                      NaN   
75%                    8.000000                      NaN   
max                  297.000000                      NaN   

       Nombre de trains en retard au départ  \
count                           9598.000000   
mean                              87.775891   
std                               89.711554   
min                                0.000000   
25%                               22.000000   
50%                               54.000000   
75%                              128.000000   
max                              596.000000   

       Retard moyen des trains en retard au départ  \
count                                  9598.000000   
mean                                     11.597082   
std                                      11.918225   
min                                       0.000000   
25%                                       5.588875   
50%                                       9.565817   
75%                                      14.724359   
max                                     316.188095   

       Retard moyen de tous les trains au départ  \
count                                9598.000000   
mean                                    3.056784   
std                                     5.097969   
min                                  -229.269444   
25%                                     1.165647   
50%                                     2.260175   
75%                                     3.897177   
max                                    84.516667   

       Commentaire retards au départ  Nombre de trains en retard à l'arrivée  \
count                            0.0                             9598.000000   
mean                             NaN                               36.074495   
std                              NaN                               30.557789   
min                              NaN                                0.000000   
25%                              NaN                               14.000000   
50%                              NaN                               28.000000   
75%                              NaN                               49.000000   
max                              NaN                              376.000000   

       Retard moyen des trains en retard à l'arrivée  \
count                                    9598.000000   
mean                                       34.170000   
std                                        15.491757   
min                                       -40.109259   
25%                                        24.967716   
50%                                        32.599802   
75%                                        41.365968   
max                                       299.600000   

       Retard moyen de tous les trains à l'arrivée  \
count                                  9598.000000   
mean                                      5.785380   
std                                       7.469440   
min                                    -472.63888

The columns "Commentaire annulations" and "Commentaire retards au départ" are empty (and if they had containedanything, it would have been unstructured data -  comments about the cancellations and delays). Let's remove them.

I will also rename all the columns to translate them to English before going further.

##  3.1 Drop the empty columns

In [6]:
df_delay_tgv = df_delay_tgv.drop(columns=['Commentaire annulations', 'Commentaire retards au départ'])

## 3.2 Rename columns to English

In [7]:
df_delay_tgv.rename(columns={
    "Gare de départ": "Departure station",
    "Gare d'arrivée": "Arrival station",
    "Durée moyenne du trajet": "Avg trip length",
    "Nombre de circulations prévues": "Number of trips scheduled",
    "Nombre de trains annulés": "Number of trains cancelled",
    "Nombre de trains en retard au départ": "Number of trains delayed on departure",
    "Retard moyen des trains en retard au départ": "Avg delay of trains delayed on departure",
    "Retard moyen de tous les trains au départ": "Avg delay of all trains on departure",
    "Nombre de trains en retard à l'arrivée": "Number of trains delayed on arrival",
    "Retard moyen des trains en retard à l'arrivée": "Avg delay of trains delayed on arrival",
    "Retard moyen de tous les trains à l'arrivée": "Avg delay of all trains on arrival",
    "Commentaire retards à l'arrivée" : "Comments delay on arrival",
    "Nombre trains en retard > 15min": "Number of trains >15 min delay",
    "Retard moyen trains en retard > 15 (si liaison concurrencée par vol)": "Avg delay of trains >15 (if trip has a flight concurrence)",
    "Nombre trains en retard > 30min": "Number of trains >30 min delay",
    "Nombre trains en retard > 60min": "Number of trains >60 min delay",
    "Prct retard pour causes externes": "% delay from external causes",
    "Prct retard pour cause infrastructure": "% delay infrastructure cause",
    "Prct retard pour cause gestion trafic": "% delay traffic management cause",
    "Prct retard pour cause matériel roulant": "% delay rolling stock cause",
    "Prct retard pour cause gestion en gare et réutilisation de matériel": "% delay station management and reutilization of stock",
    "Prct retard pour cause prise en compte voyageurs (affluence, gestions PSH, correspondances)": "% delay because of accommodation of passengers (crowd, disability, connections)"
}, inplace = True)

In [8]:
# Explore again
df_delay_tgv.describe()

Avg trip length  Number of trips scheduled  Number of trains cancelled  \
count      9598.000000                9598.000000                 9598.000000   
mean        170.010940                 265.281413                   10.070744   
std          87.300663                 178.879185                   24.784164   
min           0.000000                   0.000000                    0.000000   
25%         100.000000                 147.000000                    0.000000   
50%         163.000000                 227.000000                    2.000000   
75%         222.000000                 346.000000                    8.000000   
max         786.000000                1100.000000                  297.000000   

       Number of trains delayed on departure  \
count                            9598.000000   
mean                               87.775891   
std                                89.711554   
min                                 0.000000   
25%                                22.000000   
50%                                54.000000   
75%                               128.000000   
max                               596.000000   

       Avg delay of trains delayed on departure  \
count                               9598.000000   
mean                                  11.597082   
std                                   11.918225   
min                                    0.000000   
25%                                    5.588875   
50%                                    9.565817   
75%                                   14.724359   
max                                  316.188095   

       Avg delay of all trains on departure  \
count                           9598.000000   
mean                               3.056784   
std                                5.097969   
min                             -229.269444   
25%                                1.165647   
50%                                2.260175   
75%                                3.897177   
max                               84.516667   

       Number of trains delayed on arrival  \
count                          9598.000000   
mean                             36.074495   
std                              30.557789   
min                               0.000000   
25%                              14.000000   
50%                              28.000000   
75%                              49.000000   
max                             376.000000   

       Avg delay of trains delayed on arrival  \
count                             9598.000000   
mean                                34.170000   
std                                 15.491757   
min                                -40.109259   
25%                                 24.967716   
50%                                 32.599802   
75%                                 41.365968   
max                                299.600000   

       Avg delay of all trains on arrival  Number of trains >15 min delay  \
count                         9598.000000                     9598.000000   
mean                             5.785380                       25.588248   
std                              7.469440                       21.927697   
min                           -472.638889                        0.000000   
25%                              3.245905                       10.000000   
50%                              5.103441                       20.000000   
75%                              7.783922                       35.000000   
max                             92.000000                      312.000000   

       Avg delay of trains >15 (if trip has a flight concurrence)  \
count                                        9598.000000            
mean                                           33.754032            
std                                            19.720504            
min                                            -4.000000            
25%                                            22.725868        

The "min" shows some negative values in several columns with delay values: it shouldn't be possible. Let's check it out.

## 3.3 Deal with outliers/impossible values

In [9]:
# Identify rows with negative values
df_delay_tgv.loc[df_delay_tgv['Avg delay of all trains on departure'] < 0]

Date        Service           Departure station      Arrival station  \
416   2018-04       National  SAINT ETIENNE CHATEAUCREUX           PARIS LYON   
478   2018-04       National                     ST MALO   PARIS MONTPARNASSE   
487   2018-04       National        MARSEILLE ST CHARLES            TOURCOING   
617   2018-05  International                      GENEVE           PARIS LYON   
654   2018-06       National          PARIS MONTPARNASSE  ST PIERRE DES CORPS   
...       ...            ...                         ...                  ...   
8098  2023-06       National                   TOURCOING     BORDEAUX ST JEAN   
8492  2023-09  International                  PARIS LYON               ITALIE   
9306  2024-04       National                     ST MALO   PARIS MONTPARNASSE   
9360  2024-05  International                  PARIS LYON            BARCELONA   
9569  2024-06       National                     ST MALO   PARIS MONTPARNASSE   

      Avg trip length  Number of trips scheduled  Number of trains cancelled  \
416               170                         81                          18   
478               159                         81                          24   
487               301                         25                           7   
617               180                        217                          59   
654                61                        392                          69   
...               ...                        ...                         ...   
8098              305                         30                           0   
8492              423                         30                          16   
9306              153                         95                           0   
9360              409                         64                           0   
9569              153                         95                           0   

      Number of trains delayed on departure  \
416                                       8   
478                                      11   
487                                       0   
617                                       8   
654                                      19   
...                                     ...   
8098                                      0   
8492                                      0   
9306                                      1   
9360                                      6   
9569                                      3   

      Avg delay of trains delayed on departure  \
416                                   4.945833   
478                                   3.584848   
487                                   0.000000   
617                                  14.875000   
654                                  22.894737   
...                                        ...   
8098                                  0.000000   
8492                                  0.000000   
9306                                  1.533333   
9360                                  1.494444   
9569                                  3.061111   

      Avg delay of all trains on departure  \
416                              -1.066667   
478                              -2.158187   
487                              -0.370370   
617                              -1.759494   
654                              -0.069247   
...                                    ...   
8098                             -0.004598   
8492                             -0.029762   
9306                             -0.179298   
9360                             -0.166667   
9569                             -0.085263   

      Number of trains delayed on arrival  \
416                                     5   
478                                    17   
487                                     0   
617                                    43   
654                                    64   
...                                   ...   
8098                                    2

### Deal with "Avg delay of all trains on departure" column

The column "Avg delay of trains delayed on departure" doesn't have negative values, so we can use it to calculate the value of the column "Average delay of all trains on departure", using these formulas:

* Avg delay of trains delayed on departure = Sum of delays at departure / Number of trains delayed on departure

* Avg delay of all trains on departure = Sum of delays on departure / Number of trips scheduled

* ##### => Avg delay of all trains on departure = Avg delay of trains delayed on departure * Number of trains delayed on departure / Number of trips scheduled

With the exception where the Number of trips scheduled is 0 ; then the avg delay has to be 0.

In [10]:
# Calculate the replacement values
replacement_values = np.where(
    df_delay_tgv["Number of trips scheduled"] != 0,
    df_delay_tgv["Avg delay of trains delayed on departure"] *
    df_delay_tgv["Number of trains delayed on departure"] /
    df_delay_tgv["Number of trips scheduled"],
    0
)

In [11]:
# Replace the negative values in "Average delay of all trains at departure"
df_delay_tgv.loc[df_delay_tgv["Avg delay of all trains on departure"] < 0, "Avg delay of all trains on departure"] = replacement_values[df_delay_tgv["Avg delay of all trains on departure"] < 0]

In [12]:
# Check result
df_delay_tgv.loc[df_delay_tgv['Avg delay of all trains on departure'] < 0]

Empty DataFrame
Columns: [Date, Service, Departure station, Arrival station, Avg trip length, Number of trips scheduled, Number of trains cancelled, Number of trains delayed on departure, Avg delay of trains delayed on departure, Avg delay of all trains on departure, Number of trains delayed on arrival, Avg delay of trains delayed on arrival, Avg delay of all trains on arrival, Comments delay on arrival, Number of trains >15 min delay, Avg delay of trains >15 (if trip has a flight concurrence), Number of trains >30 min delay, Number of trains >60 min delay, % delay from external causes, % delay infrastructure cause, % delay traffic management cause, % delay rolling stock cause, % delay station management and reutilization of stock, % delay because of accommodation of passengers (crowd, disability, connections)]
Index: []

In [13]:
# Check result
df_delay_tgv.describe()

Avg trip length  Number of trips scheduled  Number of trains cancelled  \
count      9598.000000                9598.000000                 9598.000000   
mean        170.010940                 265.281413                   10.070744   
std          87.300663                 178.879185                   24.784164   
min           0.000000                   0.000000                    0.000000   
25%         100.000000                 147.000000                    0.000000   
50%         163.000000                 227.000000                    2.000000   
75%         222.000000                 346.000000                    8.000000   
max         786.000000                1100.000000                  297.000000   

       Number of trains delayed on departure  \
count                            9598.000000   
mean                               87.775891   
std                                89.711554   
min                                 0.000000   
25%                                22.000000   
50%                                54.000000   
75%                               128.000000   
max                               596.000000   

       Avg delay of trains delayed on departure  \
count                               9598.000000   
mean                                  11.597082   
std                                   11.918225   
min                                    0.000000   
25%                                    5.588875   
50%                                    9.565817   
75%                                   14.724359   
max                                  316.188095   

       Avg delay of all trains on departure  \
count                           9598.000000   
mean                               3.136249   
std                                4.101583   
min                                0.000000   
25%                                1.174019   
50%                                2.263914   
75%                                3.904016   
max                               84.516667   

       Number of trains delayed on arrival  \
count                          9598.000000   
mean                             36.074495   
std                              30.557789   
min                               0.000000   
25%                              14.000000   
50%                              28.000000   
75%                              49.000000   
max                             376.000000   

       Avg delay of trains delayed on arrival  \
count                             9598.000000   
mean                                34.170000   
std                                 15.491757   
min                                -40.109259   
25%                                 24.967716   
50%                                 32.599802   
75%                                 41.365968   
max                                299.600000   

       Avg delay of all trains on arrival  Number of trains >15 min delay  \
count                         9598.000000                     9598.000000   
mean                             5.785380                       25.588248   
std                              7.469440                       21.927697   
min                           -472.638889                        0.000000   
25%                              3.245905                       10.000000   
50%                              5.103441                       20.000000   
75%                              7.783922                       35.000000   
max                             92.000000                      312.000000   

       Avg delay of trains >15 (if trip has a flight concurrence)  \
count                                        9598.000000            
mean                                           33.754032            
std                                            19.720504            
min                                            -4.000000            
25%                                            22.725868        

All good, the column doesn't have negative values anymore.

### Deal with "Avg delay of trains delayed on arrival" column

In [14]:
df_delay_tgv.loc[df_delay_tgv['Avg delay of trains delayed on arrival'] < 0]

Date   Service Departure station Arrival station  Avg trip length  \
2886  2019-11  National       MONTPELLIER      PARIS LYON              380   
2889  2019-11  National             NIMES      PARIS LYON              224   

      Number of trips scheduled  Number of trains cancelled  \
2886                        227                          11   
2889                        226                          11   

      Number of trains delayed on departure  \
2886                                    189   
2889                                    190   

      Avg delay of trains delayed on departure  \
2886                                  4.910406   
2889                                  8.765614   

      Avg delay of all trains on departure  \
2886                              4.277469   
2889                              7.369322   

      Number of trains delayed on arrival  \
2886                                   44   
2889                                   63   

      Avg delay of trains delayed on arrival  \
2886                              -30.512500   
2889                              -40.109259   

      Avg delay of all trains on arrival Comments delay on arrival  \
2886                         -150.562114                       NaN   
2889                         -151.291008                       NaN   

      Number of trains >15 min delay  \
2886                              44   
2889                              44   

      Avg delay of trains >15 (if trip has a flight concurrence)  \
2886                                          34.677381            
2889                                          34.677381            

      Number of trains >30 min delay  Number of trains >60 min delay  \
2886                              18                               3   
2889                              18                               3   

      % delay from external causes  % delay infrastructure cause  \
2886                     52.272727                     13.636364   
2889                     46.774194                     17.741935   

      % delay traffic management cause  % delay rolling stock cause  \
2886                         15.909091                    15.909091   
2889                         14.516129                    12.903226   

      % delay station management and reutilization of stock  \
2886                                           2.272727       
2889                                           3.225806       

      % delay because of accommodation of passengers (crowd, disability, connections)  
2886                                            0.00000                                
2889                                            4.83871

There are only 2 negative values - these are two lines with Paris Lyon as arrival station, in Nov 2019. The absolute values of those negative numbers are close enough to the mean of the column (34.17), which suggests that the error might only be the minus sign in front of the number. Let's replace them with their absolute values.

In [15]:
df_delay_tgv.loc[df_delay_tgv["Avg delay of trains delayed on arrival"] < 0, "Avg delay of trains delayed on arrival"] = df_delay_tgv['Avg delay of trains delayed on arrival'].abs()

In [16]:
# Check result
df_delay_tgv.loc[df_delay_tgv['Avg delay of trains delayed on arrival'] < 0]

Empty DataFrame
Columns: [Date, Service, Departure station, Arrival station, Avg trip length, Number of trips scheduled, Number of trains cancelled, Number of trains delayed on departure, Avg delay of trains delayed on departure, Avg delay of all trains on departure, Number of trains delayed on arrival, Avg delay of trains delayed on arrival, Avg delay of all trains on arrival, Comments delay on arrival, Number of trains >15 min delay, Avg delay of trains >15 (if trip has a flight concurrence), Number of trains >30 min delay, Number of trains >60 min delay, % delay from external causes, % delay infrastructure cause, % delay traffic management cause, % delay rolling stock cause, % delay station management and reutilization of stock, % delay because of accommodation of passengers (crowd, disability, connections)]
Index: []

All good, the column doesn't have negative values anymore.

### Deal with "Avg delay of all trains on arrival" column

In [17]:
# Observe the negative values
df_delay_tgv.loc[df_delay_tgv['Avg delay of all trains on arrival'] < 0]

Date        Service     Departure station    Arrival station  \
419   2018-04  International  MARSEILLE ST CHARLES             MADRID   
505   2018-04       National       PARIS VAUGIRARD   BORDEAUX ST JEAN   
1074  2018-09       National    PARIS MONTPARNASSE            ST MALO   
1099  2018-09       National                RENNES    PARIS VAUGIRARD   
1739  2019-02  International  MARSEILLE ST CHARLES             MADRID   
...       ...            ...                   ...                ...   
8255  2023-07       National             TOURCOING   BORDEAUX ST JEAN   
8318  2023-08       National             TOURCOING   BORDEAUX ST JEAN   
8973  2024-01  International             PARIS EST          FRANCFORT   
9013  2024-02       National    PARIS MONTPARNASSE  TOULOUSE MATABIAU   
9426  2024-05       National            STRASBOURG             NANTES   

      Avg trip length  Number of trips scheduled  Number of trains cancelled  \
419               464                         23                          16   
505               165                         57                          15   
1074              171                         97                           0   
1099              116                         27                           0   
1739              464                         28                           0   
...               ...                        ...                         ...   
8255              304                         29                           0   
8318              304                         28                           0   
8973              258                        120                          58   
9013              271                        207                          11   
9426              335                         57                           0   

      Number of trains delayed on departure  \
419                                       2   
505                                       4   
1074                                      4   
1099                                      2   
1739                                     10   
...                                     ...   
8255                                      4   
8318                                      1   
8973                                      5   
9013                                     23   
9426                                      7   

      Avg delay of trains delayed on departure  \
419                                   1.208333   
505                                   6.733333   
1074                                  4.279167   
1099                                  2.275000   
1739                                  2.031667   
...                                        ...   
8255                                  2.750000   
8318                                  3.000000   
8973                                  3.743333   
9013                                 10.257246   
9426                                  4.700000   

      Avg delay of all trains on departure  \
419                               0.383333   
505                               0.099603   
1074                              0.176460   
1099                              0.174074   
1739                              0.916667   
...                                    ...   
8255                              0.393103   
8318                              0.107143   
8973                              0.302778   
9013                              0.753061   
9426                              0.640351   

      Number of trains delayed on arrival  \
419                                     0   
505                                     2   
1074                                    2   
1099                                    0   
1739                                    1   
...                                   ...   
8255                                    1   
8318                                    1   
8973                                    8   
90

Let's proceed in a similar fashion as the for column "Avg delay of all trains on departure". Based on the same logic:

Avg delay of all trains on arrival = Avg delay of trains delayed on arrival * Number of trains delayed on arrival / Number of trips scheduled

Unless the Number of trips scheduled = 0; then, the value is 0.

In [18]:
# Calculate the replacement values
replacement_values2 = np.where(
    df_delay_tgv["Number of trips scheduled"] != 0,
    df_delay_tgv["Avg delay of trains delayed on arrival"] *
    df_delay_tgv["Number of trains delayed on arrival"] /
    df_delay_tgv["Number of trips scheduled"],
    0
)

In [19]:
# Replace the negative values in "Average delay of all trains at departure"
df_delay_tgv.loc[df_delay_tgv["Avg delay of all trains on arrival"] < 0, "Avg delay of all trains on arrival"] = replacement_values2[df_delay_tgv["Avg delay of all trains on arrival"] < 0]

In [20]:
# Check result
df_delay_tgv.describe()

Avg trip length  Number of trips scheduled  Number of trains cancelled  \
count      9598.000000                9598.000000                 9598.000000   
mean        170.010940                 265.281413                   10.070744   
std          87.300663                 178.879185                   24.784164   
min           0.000000                   0.000000                    0.000000   
25%         100.000000                 147.000000                    0.000000   
50%         163.000000                 227.000000                    2.000000   
75%         222.000000                 346.000000                    8.000000   
max         786.000000                1100.000000                  297.000000   

       Number of trains delayed on departure  \
count                            9598.000000   
mean                               87.775891   
std                                89.711554   
min                                 0.000000   
25%                                22.000000   
50%                                54.000000   
75%                               128.000000   
max                               596.000000   

       Avg delay of trains delayed on departure  \
count                               9598.000000   
mean                                  11.597082   
std                                   11.918225   
min                                    0.000000   
25%                                    5.588875   
50%                                    9.565817   
75%                                   14.724359   
max                                  316.188095   

       Avg delay of all trains on departure  \
count                           9598.000000   
mean                               3.136249   
std                                4.101583   
min                                0.000000   
25%                                1.174019   
50%                                2.263914   
75%                                3.904016   
max                               84.516667   

       Number of trains delayed on arrival  \
count                          9598.000000   
mean                             36.074495   
std                              30.557789   
min                               0.000000   
25%                              14.000000   
50%                              28.000000   
75%                              49.000000   
max                             376.000000   

       Avg delay of trains delayed on arrival  \
count                             9598.000000   
mean                                34.184716   
std                                 15.459253   
min                                  0.000000   
25%                                 24.984962   
50%                                 32.601505   
75%                                 41.365968   
max                                299.600000   

       Avg delay of all trains on arrival  Number of trains >15 min delay  \
count                         9598.000000                     9598.000000   
mean                             5.961176                       25.588248   
std                              4.151375                       21.927697   
min                              0.000000                        0.000000   
25%                              3.257813                       10.000000   
50%                              5.110889                       20.000000   
75%                              7.791706                       35.000000   
max                             92.000000                      312.000000   

       Avg delay of trains >15 (if trip has a flight concurrence)  \
count                                        9598.000000            
mean                                           33.754032            
std                                            19.720504            
min                                            -4.000000            
25%                                            22.725868        

All good, the column doesn't have negative values anymore.

#### Deal with "Avg delay of trains >15 (if trip has a flight concurrence)" column

This column is actually not supposed to have any value under 15.

In [21]:
# Locate values under 15
df_delay_tgv.loc[df_delay_tgv['Avg delay of trains >15 (if trip has a flight concurrence)'] < 15]

Date        Service   Departure station      Arrival station  \
0     2018-01       National    BORDEAUX ST JEAN   PARIS MONTPARNASSE   
1     2018-01       National   LA ROCHELLE VILLE   PARIS MONTPARNASSE   
2     2018-01       National  PARIS MONTPARNASSE              QUIMPER   
3     2018-01       National  PARIS MONTPARNASSE              ST MALO   
4     2018-01       National  PARIS MONTPARNASSE  ST PIERRE DES CORPS   
...       ...            ...                 ...                  ...   
9333  2024-04  International          PARIS LYON               ITALIE   
9360  2024-05  International          PARIS LYON            BARCELONA   
9426  2024-05       National          STRASBOURG               NANTES   
9432  2024-05  International          PARIS LYON               ITALIE   
9546  2024-06  International          PARIS LYON            BARCELONA   

      Avg trip length  Number of trips scheduled  Number of trains cancelled  \
0                 141                        870                           5   
1                 165                        222                           0   
2                 220                        248                           1   
3                 156                        102                           0   
4                  61                        391                           2   
...               ...                        ...                         ...   
9333              502                         28                           0   
9360              409                         64                           0   
9426              335                         57                           0   
9432              502                         18                           0   
9546              409                         70                           0   

      Number of trains delayed on departure  \
0                                       289   
1                                         8   
2                                        37   
3                                        12   
4                                        61   
...                                     ...   
9333                                      1   
9360                                      6   
9426                                      7   
9432                                      2   
9546                                      8   

      Avg delay of trains delayed on departure  \
0                                    11.247809   
1                                     2.875000   
2                                     9.501351   
3                                    19.912500   
4                                     7.796995   
...                                        ...   
9333                                 14.766667   
9360                                  1.494444   
9426                                  4.700000   
9432                                 10.741667   
9546                                 10.039583   

      Avg delay of all trains on departure  \
0                                 3.693179   
1                                 0.095796   
2                                 1.003981   
3                                 1.966667   
4                                 0.886889   
...                                    ...   
9333                              0.459524   
9360                              0.140104   
9426                              0.640351   
9432                              1.194444   
9546                              0.878095   

      Number of trains delayed on arrival  \
0                                     147   
1                                      34   
2                                      26   
3                                      13   
4                                      71   
...                                   ...   
9333                                    0   
9360                                    0   
9426                                    0   
94

2157 rows with an incoherent value out of the 9598 rows of the data set: that's 22%. This is way too much to impute values or delete the rows without affecting the analysis. This variable is not very interesting, it is out of the scope of our topic: let's simply delete the column.

In [22]:
df_delay_tgv = df_delay_tgv.drop(columns = ['Avg delay of trains >15 (if trip has a flight concurrence)'])

In [23]:
# Check result
df_delay_tgv.describe()

Avg trip length  Number of trips scheduled  Number of trains cancelled  \
count      9598.000000                9598.000000                 9598.000000   
mean        170.010940                 265.281413                   10.070744   
std          87.300663                 178.879185                   24.784164   
min           0.000000                   0.000000                    0.000000   
25%         100.000000                 147.000000                    0.000000   
50%         163.000000                 227.000000                    2.000000   
75%         222.000000                 346.000000                    8.000000   
max         786.000000                1100.000000                  297.000000   

       Number of trains delayed on departure  \
count                            9598.000000   
mean                               87.775891   
std                                89.711554   
min                                 0.000000   
25%                                22.000000   
50%                                54.000000   
75%                               128.000000   
max                               596.000000   

       Avg delay of trains delayed on departure  \
count                               9598.000000   
mean                                  11.597082   
std                                   11.918225   
min                                    0.000000   
25%                                    5.588875   
50%                                    9.565817   
75%                                   14.724359   
max                                  316.188095   

       Avg delay of all trains on departure  \
count                           9598.000000   
mean                               3.136249   
std                                4.101583   
min                                0.000000   
25%                                1.174019   
50%                                2.263914   
75%                                3.904016   
max                               84.516667   

       Number of trains delayed on arrival  \
count                          9598.000000   
mean                             36.074495   
std                              30.557789   
min                               0.000000   
25%                              14.000000   
50%                              28.000000   
75%                              49.000000   
max                             376.000000   

       Avg delay of trains delayed on arrival  \
count                             9598.000000   
mean                                34.184716   
std                                 15.459253   
min                                  0.000000   
25%                                 24.984962   
50%                                 32.601505   
75%                                 41.365968   
max                                299.600000   

       Avg delay of all trains on arrival  Number of trains >15 min delay  \
count                         9598.000000                     9598.000000   
mean                             5.961176                       25.588248   
std                              4.151375                       21.927697   
min                              0.000000                        0.000000   
25%                              3.257813                       10.000000   
50%                              5.110889                       20.000000   
75%                              7.791706                       35.000000   
max                             92.000000                      312.000000   

       Number of trains >30 min delay  Number of trains >60 min delay  \
count                     9598.000000                     9598.000000   
mean                        12.169098                        4.459158   
std                         11.440176                        5.079933   
min                          0.000000                        0.000000   
25%                          4.000000       

The column was successfully dropped.

### Deal with "Avg trip length" column

The column "Avg trip length" should never take the value 0, but it is its minimum value. Let's have a look:

In [24]:
pd.set_option('display.max_rows', None)
df_delay_tgv.loc[df_delay_tgv['Avg trip length'] == 0]

Date        Service           Departure station  \
3406  2020-04       National          PARIS MONTPARNASSE   
3413  2020-04       National             MARNE LA VALLEE   
3414  2020-04       National             MARNE LA VALLEE   
3415  2020-04       National        MARSEILLE ST CHARLES   
3416  2020-04       National                  PARIS LYON   
3417  2020-04       National                      ANNECY   
3418  2020-04       National                  PARIS LYON   
3423  2020-04       National                  PARIS LYON   
3424  2020-04       National                      TOULON   
3427  2020-04  International                   FRANCFORT   
3429  2020-04  International                   BARCELONA   
3432  2020-04       National           LA ROCHELLE VILLE   
3436  2020-04       National          PARIS MONTPARNASSE   
3437  2020-04       National                       TOURS   
3438  2020-04       National              LYON PART DIEU   
3440  2020-04       National                  PARIS NORD   
3441  2020-04       National                   DUNKERQUE   
3442  2020-04       National        MARSEILLE ST CHARLES   
3443  2020-04       National        MARSEILLE ST CHARLES   
3445  2020-04       National            BELLEGARDE (AIN)   
3447  2020-04       National                  PARIS LYON   
3451  2020-04       National                  PARIS LYON   
3457  2020-04       National          PARIS MONTPARNASSE   
3466  2020-04       National                  STRASBOURG   
3467  2020-04       National                  PARIS NORD   
3469  2020-04       National                       LILLE   
3473  2020-04  International                   PARIS EST   
3474  2020-04  International                  PARIS LYON   
3475  2020-04  International                      GENEVE   
3476  2020-04  International                  PARIS LYON   
3479  2020-04       National                     ST MALO   
3487  2020-04       National              LYON PART DIEU   
3488  2020-04       National                  PARIS LYON   
3492  2020-04       National                  PARIS LYON   
3493  2020-04       National   CHAMBERY CHALLES LES EAUX   
3500  2020-04  International                  PARIS LYON   
3506  2020-04       National                      RENNES   
3509  2020-04       National                      NANTES   
3511  2020-04       National            BORDEAUX ST JEAN   
3512  2020-04       National                   TOURCOING   
3514  2020-04       National                    GRENOBLE   
3516  2020-04       National                  PARIS LYON   
3517  2020-04       National                  NICE VILLE   
3519  2020-04       National                   PERPIGNAN   
3520  2020-04       National  SAINT ETIENNE CHATEAUCREUX   
3521  2020-04       National                  PARIS LYON   
3523  2020-04  International                      ITALIE   
3524  2020-04  International                    LAUSANNE   
3526  2020-04  International                      MADRID   
3527  2020-04  International        MARSEILLE ST CHARLES   
3552  2020-05  International                      MADRID   
3567  2020-05       National        MARSEILLE ST CHARLES   
3604  2020-05       National                  STRASBOURG   
3605  2020-05       National                      NANTES   
3621  2020-05       National                  PARIS LYON   
3624  2020-05  International        MARSEILLE ST CHARLES   
3635  2020-05       National            BORDEAUX ST JEAN   
3636  2020-05       National                   TOURCOING   
3642  2020-05       National  SAINT ETIENNE CHATEAUCREUX   
3664  2020-06  International                   PARIS EST   
3690  2020-06  International                   STUTTGART   
3692  2020-06  International                      MADRID   
3740  2020-06  International        MARSEILLE ST CHARLES   
3966  2020-08  International        MARSEILLE ST CHARLES   
3991  2020-08  International                      MADRID   
4042  2020-09  International                     

All of those rows are from 2020, mostly April and May: it's probably due to Covid-19 pandemic disturbance. For all the rows, the value of "Number of trips scheduled" is 0, but "Number of trains cancelled" is mostly not 0. Let's look at the "0" in "Number of trips scheduled".

In [25]:
pd.set_option('display.max_rows', 10)
df_delay_tgv.loc[df_delay_tgv['Number of trips scheduled'] == 0]

Date        Service     Departure station       Arrival station  \
3406  2020-04       National    PARIS MONTPARNASSE               ST MALO   
3413  2020-04       National       MARNE LA VALLEE        LYON PART DIEU   
3414  2020-04       National       MARNE LA VALLEE  MARSEILLE ST CHARLES   
3415  2020-04       National  MARSEILLE ST CHARLES       MARNE LA VALLEE   
3416  2020-04       National            PARIS LYON                ANNECY   
...       ...            ...                   ...                   ...   
4235  2020-10  International  MARSEILLE ST CHARLES                MADRID   
4320  2020-11  International                MADRID  MARSEILLE ST CHARLES   
4321  2020-11  International  MARSEILLE ST CHARLES                MADRID   
4432  2020-12  International                MADRID  MARSEILLE ST CHARLES   
4510  2020-12  International  MARSEILLE ST CHARLES                MADRID   

      Avg trip length  Number of trips scheduled  Number of trains cancelled  \
3406                0                          0                           8   
3413                0                          0                          21   
3414                0                          0                          14   
3415                0                          0                          13   
3416                0                          0                           7   
...               ...                        ...                         ...   
4235                0                          0                           0   
4320                0                          0                           0   
4321                0                          0                           0   
4432                0                          0                           0   
4510                0                          0                           0   

      Number of trains delayed on departure  \
3406                                      0   
3413                                      0   
3414                                      0   
3415                                      0   
3416                                      0   
...                                     ...   
4235                                      0   
4320                                      0   
4321                                      0   
4432                                      0   
4510                                      0   

      Avg delay of trains delayed on departure  \
3406                                       0.0   
3413                                       0.0   
3414                                       0.0   
3415                                       0.0   
3416                                       0.0   
...                                        ...   
4235                                       0.0   
4320                                       0.0   
4321                                       0.0   
4432                                       0.0   
4510                                       0.0   

      Avg delay of all trains on departure  \
3406                                   0.0   
3413                                   0.0   
3414                                   0.0   
3415                                   0.0   
3416                                   0.0   
...                                    ...   
4235                                   0.0   
4320                                   0.0   
4321                                   0.0   
4432                                   0.0   
4510                                   0.0   

      Number of trains delayed on arrival  \
3406                                    0   
3413                                    0   
3414                                    0   
3415                                    0   
3416                                    0   
...                                   ...   
4235                                    0   
4320                                    0   
4321       

These are the very same rows. The issue, then, is that the "Number of train cancelled" is not always 0 and is higher than the number of trips scheduled, which doesn't make sense. Let's change the value of "Number of trains cancelled" to 0.

In [26]:
df_delay_tgv.loc[df_delay_tgv['Number of trips scheduled'] == 0, 'Number of trains cancelled'] = 0

## 3.4 Missing values

In [27]:
pd.set_option('display.max_rows', None)
df_delay_tgv.isnull().sum()

Date                                                                                  0
Service                                                                               0
Departure station                                                                     0
Arrival station                                                                       0
Avg trip length                                                                       0
Number of trips scheduled                                                             0
Number of trains cancelled                                                            0
Number of trains delayed on departure                                                 0
Avg delay of trains delayed on departure                                              0
Avg delay of all trains on departure                                                  0
Number of trains delayed on arrival                                                   0
Avg delay of trains delayed on a

The only column with missing values is "Comments delay at arrival", which is fine: it's a column of unstructured data anyway. I prefer not to get rid of the column because it might give us some insights later, if I get some surprising results or outliers in my analysis.

## 3.5 Duplicates

In [28]:
#view duplicates
df_delay_tgv[df_delay_tgv.duplicated()]

Empty DataFrame
Columns: [Date, Service, Departure station, Arrival station, Avg trip length, Number of trips scheduled, Number of trains cancelled, Number of trains delayed on departure, Avg delay of trains delayed on departure, Avg delay of all trains on departure, Number of trains delayed on arrival, Avg delay of trains delayed on arrival, Avg delay of all trains on arrival, Comments delay on arrival, Number of trains >15 min delay, Number of trains >30 min delay, Number of trains >60 min delay, % delay from external causes, % delay infrastructure cause, % delay traffic management cause, % delay rolling stock cause, % delay station management and reutilization of stock, % delay because of accommodation of passengers (crowd, disability, connections)]
Index: []

No duplicates.

## 3.6 Mixed-types columns

In [29]:
for col in df_delay_tgv.columns.tolist():
  weird = (df_delay_tgv[[col]].applymap(type) != df_delay_tgv[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_delay_tgv[weird]) > 0:
    print (col)

Comments delay on arrival


The column "Comments delay at arrival" is mixed-type: let's change it all to string.

In [30]:
df_delay_tgv['Comments delay on arrival'] = df_delay_tgv['Comments delay on arrival'].astype('str')

## 3.7 Export cleaned TGV data set

In [31]:
df_delay_tgv.to_csv(os.path.join(path, '02 Data','Prepared Data', 'delay_tgv_clean.csv'))

# 4. Explore and clean Number of travelers data set

In [32]:
df_nb_travelers.head()

Nom de la gare  Code UIC  Code postal Segmentation DRG  \
0                Abancourt  87313759        60220                C   
1                     Agde  87781278        34300                B   
2                     Agen  87586008        47000                A   
3                   Agonac  87595157        24460                C   
4  Aguilcourt - Variscourt  87171702         2190                C   

   Total Voyageurs 2023  Total Voyageurs + Non voyageurs 2023  \
0                 77974                                 77974   
1                757491                                946864   
2               1640325                               2050407   
3                  9537                                  9537   
4                 13464                                 13464   

   Total Voyageurs 2022  Total Voyageurs + Non voyageurs 2022  \
0                 71517                                 71517   
1                689202                                861503   
2               1540511                               1925639   
3                  6468                                  6468   
4                 13197                                 13197   

   Total Voyageurs 2021  Total Voyageurs + Non voyageurs 2021  \
0                 51811                                 51811   
1                561160                                701450   
2               1184007                               1480009   
3                  4119                                  4119   
4                 12294                                 12294   

   Total Voyageurs 2020  Total Voyageurs + Non voyageurs 2020  \
0                 32396                                 32396   
1                394380                                492975   
2                860964                               1076205   
3                  3271                                  3271   
4                 10436                                 10436   

   Total Voyageurs 2019  Total Voyageurs + Non voyageurs 2019  \
0                 42685                                 42685   
1                542288                                677860   
2               1211323                               1514154   
3                  2538                                  2538   
4                  8796                                  8796   

   Total Voyageurs 2018  Total Voyageurs + Non voyageurs 2018  \
0                 40228                                 40228   
1                588297                                735372   
2               1109199                               1386499   
3                  1492                                  1492   
4                  6155                                  6155   

   Total Voyageurs 2017  Total Voyageurs + Non voyageurs 2017  \
0                 43760                                 43760   
1                697091                                871364   
2               1194455                               1493068   
3                  1583                                  1583   
4                  7368                                  7368   

   Total Voyageurs 2016  Total Voyageurs + Non voyageurs 2016  \
0                 41096                                 41096   
1                660656                                825820   
2               1141620                               1427026   
3                  1134                                  1134   
4                  8979                                  8979   

   Total Voyageurs 2015  Total Voyageurs + Non voyageurs 2015  
0                 39720                                 39720  
1                662516                                828146  
2               1183150                               1478938  
3                  1127                                  1127  
4                  9821                                  9821

## 4.1 Drop useless columns

For each year, we have a column for the number of travelers and the number of travelers + visitors who didn't travel. We don't need the latest. We don't need the postal code either, nor the Segmentation DRG. Let's trop all of that. I'll keep "Code IUC", which is an unique identifier for train stations, because some train stations might have the same name and we'll need to distinguish them.

In [33]:
df_nb_travelers = df_nb_travelers.drop(columns=['Code postal', 'Segmentation DRG', 'Total Voyageurs + Non voyageurs 2023', 'Total Voyageurs + Non voyageurs 2022', 'Total Voyageurs + Non voyageurs 2021', 'Total Voyageurs + Non voyageurs 2020', 'Total Voyageurs + Non voyageurs 2019', 'Total Voyageurs + Non voyageurs 2018', 'Total Voyageurs + Non voyageurs 2017', 'Total Voyageurs + Non voyageurs 2016', 'Total Voyageurs + Non voyageurs 2015'])

In [34]:
df_nb_travelers.head(1)

Nom de la gare  Code UIC  Total Voyageurs 2023  Total Voyageurs 2022  \
0      Abancourt  87313759                 77974                 71517   

   Total Voyageurs 2021  Total Voyageurs 2020  Total Voyageurs 2019  \
0                 51811                 32396                 42685   

   Total Voyageurs 2018  Total Voyageurs 2017  Total Voyageurs 2016  \
0                 40228                 43760                 41096   

   Total Voyageurs 2015  
0                 39720

## 4.2 Rename columns to English

In [35]:
df_nb_travelers.rename(columns={
    "Nom de la gare": "Station name",
    "Code UIC": "UIC code",
    "Total Voyageurs 2023": "Total travelers 2023",
    "Total Voyageurs 2022": "Total travelers 2022",
    "Total Voyageurs 2021": "Total travelers 2021",
    "Total Voyageurs 2020": "Total travelers 2020",
    "Total Voyageurs 2019": "Total travelers 2019",
    "Total Voyageurs 2018": "Total travelers 2018",
    "Total Voyageurs 2017": "Total travelers 2017",
    "Total Voyageurs 2016": "Total travelers 2016",
    "Total Voyageurs 2015": "Total travelers 2015",
}, inplace = True)

## 4.3 Deal with outliers/impossible values

In [36]:
df_nb_travelers.describe()

UIC code  Total travelers 2023  Total travelers 2022  \
count  3.010000e+03          3.010000e+03          3.010000e+03   
mean   8.748935e+07          9.281000e+05          8.656592e+05   
std    2.067503e+05          5.795150e+06          5.438984e+06   
min    8.700970e+07          0.000000e+00          0.000000e+00   
25%    8.731325e+07          1.127850e+04          1.024800e+04   
50%    8.748726e+07          5.862850e+04          5.220250e+04   
75%    8.768411e+07          2.921128e+05          2.630058e+05   
max    8.798872e+07          2.267685e+08          2.116985e+08   

       Total travelers 2021  Total travelers 2020  Total travelers 2019  \
count          3.010000e+03          3.010000e+03          3.010000e+03   
mean           6.908668e+05          4.760994e+05          9.104975e+05   
std            3.982991e+06          3.227785e+06          6.096153e+06   
min            0.000000e+00          0.000000e+00          0.000000e+00   
25%            7.463000e+03          5.528250e+03          7.721000e+03   
50%            3.857650e+04          3.061650e+04          4.436550e+04   
75%            2.054730e+05          1.562675e+05          2.384548e+05   
max            1.266210e+08          1.144684e+08          2.447444e+08   

       Total travelers 2018  Total travelers 2017  Total travelers 2016  \
count          3.010000e+03          3.010000e+03          3.010000e+03   
mean           8.934209e+05          9.097895e+05          8.826474e+05   
std            6.078375e+06          6.158213e+06          6.018658e+06   
min            0.000000e+00          0.000000e+00          0.000000e+00   
25%            6.801750e+03          7.503750e+03          7.645000e+03   
50%            4.024950e+04          4.338750e+04          4.170700e+04   
75%            2.199030e+05          2.334558e+05          2.235845e+05   
max            2.444656e+08          2.479632e+08          2.422731e+08   

       Total travelers 2015  
count          3.010000e+03  
mean           8.728819e+05  
std            5.938435e+06  
min            0.000000e+00  
25%            7.789750e+03  
50%            4.334500e+04  
75%            2.303582e+05  
max            2.383561e+08

The numerical values seem to make sense, except for the 0 in the min. It means a station with no travelers at all. Let's check it out.

In [37]:
pd.set_option('display.max_rows', None)
df_nb_travelers.loc[df_nb_travelers['Total travelers 2023']==0]

Station name  UIC code  Total travelers 2023  \
185                           Exideuil  87592881                     0   
221              Hymont - Mattaincourt  87144220                     0   
346                             Merrey  87142406                     0   
364                            Montbré  87171256                     0   
388                      Nonant-le-Pin  87444570                     0   
400                   Pargny-sur-Saulx  87174334                     0   
418                        Pont-Hébert  87447185                     0   
420                      Pont-du-Casse  87586404                     0   
480              Saint-Martin-Bellevue  87746222                     0   
491                        Saint-Sever  87447714                     0   
513                   Acheux - Franleu  87316745                     0   
540                   Séon Saint-Henri  87751636                     0   
608              Vignols - Saint-Solve  87594655                     0   
627                   Bolbec - Nointot  87413351                     0   
638                      Voellerdingen  87215640                     0   
652                        Wittelsheim  87182683                     0   
677                  Campagne-sur-Aude  87615229                     0   
694                           Champlan  87725457                     0   
735                   Coussac-Bonneval  87592840                     0   
788                    Glos - Montfort  87415349                     0   
884                          Louhossoa  87673673                     0   
1022                        Pontgibaud  87641316                     0   
1033                Quesnoy-le-Montant  87317487                     0   
1048                        Richwiller  87182675                     0   
1131                       Taillebourg  87491043                     0   
1157                           Valdieu  87181081                     0   
1196               Zornhoff Monswiller  87212217                     0   
1332                   Colroy - Lubine  87144386                     0   
1343                        Couliboeuf  87444232                     0   
1364                         Douzillac  87595280                     0   
1377                      Feuquerolles  87316752                     0   
1441                    La Bachellerie  87595595                     0   
1459                    Labarthe-Inard  87611129                     0   
1538                Milhac-d'Auberoche  87595538                     0   
1610                          Pommevic  87618058                     0   
1635               Roumazières-Loubert  87583468                     0   
1636                 Ruelle-sur-Touvre  87583203                     0   
1673           Saint-Pierre-du-Vauvray  87415646                     0   
1698                           Sillery  87171637                     0   
1713                Tillières-sur-Avre  87393587                     0   
1734               Verneuil-sur-Vienne  87592675                     0   
1750                 Villers sur Thère  87316422                     0   
1756             Virville - Manneville  87413336                     0   
1782                           Andilly  87142364                     0   
1835                            Bourth  87393603                     0   
1851              Carantilly - Marigny  87447144                     0   
1991             La Miouze - Rochefort  87641290                     0   
1998                  La Rochefoucauld  87583427                     0   
2042                      Les Islettes  87175711                     0   
2080                           Malause  87611871                     0   
2092               Mauzens-et-Miremont  87595728                     0   
2119                   Mulhouse Musées  87534339                     0   
2130                            Nissan  87781179                     0   
2164                       Pie

In [38]:
df_nb_travelers.loc[df_nb_travelers['Total travelers 2015']==0]

Station name  UIC code  \
5                          Aigrefeuille le Thou  87485193   
50                                Base Aérienne  87699223   
168                                       Delle  87184440   
231                         L'Etang les Sablons  87726109   
242                                 La Couronne  87583617   
276                  Le Mans Hôpital-Université  87743872   
387                                Noisy-le-Roi  87733659   
393                          Nîmes Pont du Gard  87703975   
400                            Pargny-sur-Saulx  87174334   
525                                Allée Royale  87710913   
585                            Bagnols-sur-Cèze  87764571   
622                     Villetaneuse Université  87697318   
638                               Voellerdingen  87215640   
644                                     Wambaix  87345595   
652                                 Wittelsheim  87182683   
694                                    Champlan  87725457   
812                                   Joncherey  87700153   
890                        Lurbe-Saint-Christau  87672741   
965                                      Méroux  87184408   
1018                          Pont-Saint-Esprit  87764555   
1048                                 Richwiller  87182675   
1141                        Toulon Sainte-Musse  87742262   
1157                                    Valdieu  87181081   
1196                        Zornhoff Monswiller  87212217   
1237                                     Bailly  87710780   
1408                               Grandvillars  87184432   
1453                                  La Jarrie  87485185   
1479                Le Bouscat - Saint-Germaine  87738096   
1626                                  Remoulins  87775395   
1782                                    Andilly  87142364   
1832                                   Boulazac  87725473   
1892                                  Danjoutin  87700146   
2050                            Lisière Pereire  87733683   
2089                          Marsac-sur-L'isle  87595207   
2108                  Montpellier Sud de France  87688887   
2119                            Mulhouse Musées  87534339   
2260                          Sainte-Marguerite  87287193   
2405                                Ballersdorf  87181073   
2422                                      Bidos  87691949   
2447                                  Brunstatt  87181032   
2531                   Dugny - La Courneuve T11  87697342   
2598                              Irigny Yvours  87722249   
2686                    Les Portes de Saint-Cyr  87710921   
2728                               Massy Europe  87725465   
2767                                 Morvillars  87184424   
2772               Mulhouse Zu Rhein Tram-Train  87472639   
2851                                 Rosa Parks  87654798   
2883  Saint-Germain-en-Laye Bel-Air - Fourqueux  87733675   
2929                     Stains La Cerisaie T11  87697334   
2954                                    Trélazé  87701532   

      Total travelers 2023  Total travelers 2022  Total travelers 2021  \
5                    61127                 51200                 32989   
50                    2714                  2079                   659   
168                  32269                 31341                 19250   
231                 148199                 70338                     0   
242                    364                     0                     0   
276                  10061                     0                     0   
387                 354395                168105                     0   
393                1054535                745834                590063   
400                      0                     0                     0   
525                  46010                 21783                     0   
585                  43039                 18495                     0   
622                4152466         

73 stations with 0 travelers in 2023: most of them still had travelers in 2015 and seem to have closed in the following years.

46 stations with no travelers in 2015: most of them start having travelers in the folowing years, they must be new stations.

Few stations with 0 travelers recorded ever. They will not disturb our analysis.

No action taken.

## 4.4 Missing values

In [39]:
df_nb_travelers.isnull().sum()

Station name            0
UIC code                0
Total travelers 2023    0
Total travelers 2022    0
Total travelers 2021    0
Total travelers 2020    0
Total travelers 2019    0
Total travelers 2018    0
Total travelers 2017    0
Total travelers 2016    0
Total travelers 2015    0
dtype: int64

No missing values.

## 4.5 Duplicates

In [40]:
# View duplicates in data set
df_nb_travelers[df_nb_travelers.duplicated()]

Empty DataFrame
Columns: [Station name, UIC code, Total travelers 2023, Total travelers 2022, Total travelers 2021, Total travelers 2020, Total travelers 2019, Total travelers 2018, Total travelers 2017, Total travelers 2016, Total travelers 2015]
Index: []

No duplicates.

In [41]:
# View duplicates in Station name column
df_nb_travelers[df_nb_travelers['Station name'].duplicated()]

Station name  UIC code  Total travelers 2023  Total travelers 2022  \
1813   Belleville  87141804                 31805                 26977   

      Total travelers 2021  Total travelers 2020  Total travelers 2019  \
1813                 20628                 17494                 25006   

      Total travelers 2018  Total travelers 2017  Total travelers 2016  \
1813                 23040                 28306                 28504   

      Total travelers 2015  
1813                 28800

In [42]:
df_nb_travelers.loc[df_nb_travelers['Station name'] == 'Belleville']

Station name  UIC code  Total travelers 2023  Total travelers 2022  \
1254   Belleville  87486142                134484                113138   
1813   Belleville  87141804                 31805                 26977   

      Total travelers 2021  Total travelers 2020  Total travelers 2019  \
1254                 71832                 57289                 85566   
1813                 20628                 17494                 25006   

      Total travelers 2018  Total travelers 2017  Total travelers 2016  \
1254                 73772                 67909                 58233   
1813                 23040                 28306                 28504   

      Total travelers 2015  
1254                 49298  
1813                 28800

Indeed, there are two stations with the same name, but as we can distinguisd them with the UIC code, it's alright. 

## 4.6 Mixed-types columns

In [43]:
for col in df_nb_travelers.columns.tolist():
  weird = (df_nb_travelers[[col]].applymap(type) != df_nb_travelers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_nb_travelers[weird]) > 0:
    print (col)

No mixed-types columns.

## 4.7 Export Number of travelers data set

In [44]:
df_nb_travelers.to_csv(os.path.join(path, '02 Data','Prepared Data', 'df_nb_travelers_clean.csv'))